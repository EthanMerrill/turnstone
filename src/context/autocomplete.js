import React, { createContext, useState, useEffect } from 'react'

const AutocompleteContext = createContext()

const AutocompleteContextProvider = (props) => {
  const { children, splitChar, styles = {}, text = '' } = props
  const [queryState, setQueryState] = useState(text)
  const [highlightedState, setHighlightedState] = useState()
  const [selectedState, setSelectedState] = useState()
  const [customStyles] = useState(styles)
  const [splitCharState] = useState(splitChar)

  useEffect(() => setQueryState(text), [text])

  return (
    <AutocompleteContext.Provider
      value={{
        queryState,
        setQueryState,
        highlightedState,
        setHighlightedState,
        selectedState,
        setSelectedState,
        customStyles,
        splitCharState
      }}>
      {children}
    </AutocompleteContext.Provider>
  )
}

export { AutocompleteContext, AutocompleteContextProvider }
