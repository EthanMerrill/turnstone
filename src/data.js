export const fruits = [
  "Apple",
  "Apricot",
  "Avocado",
  "Banana",
  "Bartlett pear",
  "Bilberry",
  "Blackberry",
  "Blackcap",
  "Blackcurrant",
  "Blueberry",
  "Cantaloupe",
  "Cherry",
  "Cherimoya",
  "Clementine",
  "Currant",
  "Date",
  "Damson",
  "Durian",
  "Eggplant",
  "Elderberry",
  "Feijoa",
  "Gooseberry",
  "Grape",
  "Grapefruit",
  "Guava",
  "Honeydew melon",
  "Huckleberry",
  "Jackfruit",
  "Jambul",
  "Kiwi fruit",
  "Kumquat",
  "Legume",
  "Lemon",
  "Lime",
  "Lychee",
  "Mango",
  "Mangostine",
  "Melon",
  "Rock melon",
  "Nectarine",
  "Orange",
  "Peach",
  "Pear",
  "Physalis",
  "Pineapple",
  "Pitaya",
  "Plum",
  "Pomegranate",
  "Prune",
  "Raisin",
  "Raspberry",
  "Rambutan",
  "Redcurrant",
  "Salal berry",
  "Satsuma",
  "Star fruit",
  "Strawberry",
  "Tangerine",
  "Tomato",
  "Ugli fruit",
  "Watermelon",
  "Western raspberry",
  "Williams pear",
  "Ziziphus mauritiana"
]

export const vegetables = [
  "Acorn Squash",
  "Alfalfa Sprouts",
  "Artichoke",
  "Arugula",
  "Asparagus",
  "Aubergine",
  "Azuki Beans",
  "Banana Squash",
  "Bean Sprouts",
  "Beet Greens",
  "Beetroot",
  "Bell Pepper",
  "Bitter Melon",
  "Black-Eyed Peas",
  "Black Beans",
  "Bok Choy",
  "Borlotti Bean",
  "Broad Beans",
  "Broccoflower (A Hybrid)",
  "Broccoli (Calabrese)",
  "Brussels Sprouts",
  "Butter Bean",
  "Butternut Squash",
  "Cabbage",
  "Carrot",
  "Cauliflower",
  "Ceci Beans",
  "Celeriac",
  "Celery",
  "Chard",
  "Chickpeas",
  "Chili Pepper",
  "Chives",
  "Collard Greens",
  "Corms",
  "Courgette",
  "Cucumber",
  "Daikon Or White Radish",
  "Delicata",
  "Eddoe",
  "Eggplant",
  "Endive",
  "Fennel",
  "Fiddleheads",
  "Frisee",
  "Garbanzos",
  "Garlic",
  "Gem Squash",
  "Ginger",
  "Green Beans",
  "Green Onion",
  "Greens",
  "Habanero",
  "Horseradish",
  "Hubbard Squash",
  "Jalape√±o",
  "Jerusalem Artichoke",
  "Jicama",
  "Kale",
  "Kidney Beans",
  "Kohlrabi",
  "Konjac",
  "Leek",
  "Legumes",
  "Lentils",
  "Lettuce",
  "Lima Beans",
  "Mangel-Wurzel",
  "Mange Tout",
  "Marrow",
  "Mung Beans",
  "Mustard Greens",
  "Navy Beans",
  "Nettles",
  "New Zealand Spinach",
  "Okra",
  "Onion",
  "Onions",
  "Oyster Plant",
  "Parsnip",
  "Peanuts",
  "Peas",
  "Peppers",
  "Pinto Beans",
  "Potato",
  "Radicchio",
  "Radish",
  "Red Cabbage",
  "Rhubarb",
  "Rocket",
  "Runner Beans",
  "Rutabaga",
  "Salsify",
  "Savoy Cabbage",
  "Scallion",
  "Shallot",
  "Skirret",
  "Snap Peas",
  "Soy Beans",
  "Spaghetti Squash",
  "Spinach",
  "Split Peas",
  "Spring Onion",
  "Squash",
  "Sweetcorn",
  "Sweet Potato",
  "Taro",
  "Tat Soi",
  "Tomato",
  "Topinambur",
  "Tubers",
  "Turnip",
  "Wasabi",
  "Water Chestnut",
  "Watercress",
  "Yam",
  "Zucchini"
]

export const books = [
  { title: 'Last Argument of Kings', author: 'Joe Abercrombie' },
  { title: 'Legend', author: 'Marie Lu' },
  { title: 'Life After Life', author: 'Kate Atkinson' },
  { title: 'Like Water for Chocolate', author: 'Laura Esquivel' },
  { title: 'Little House on the Prairie', author: 'Laura Ingalls Wilder' },
  { title: 'Little Women', author: 'Louisa May Alcott' },
  { title: 'Long Day\'s Journey into Night', author: 'Eugene O\'Neill' },
  { title: 'Lord Jim', author: 'Joseph Conrad' },
  { title: 'Lord of the Flies', author: 'William Golding' },
  { title: 'Lord of the Rings', author: 'J. R. R. Tolkien' }
]