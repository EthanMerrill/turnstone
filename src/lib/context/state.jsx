import React, { createContext, useReducer, useEffect } from 'react'
import reducer from '../reducers/reducer'
import {setQuery} from '../actions/actions'
import undef from '../utils/undef'

const StateContext = createContext()

const StateContextProvider = (props) => {
  const { styles = {}, text = '', items = [] } = props
  const { children, ...propsMinusChildren} = props
  const [state, dispatch] = useReducer(reducer, {
    query: text,
    items,
    itemsError: false,
    canShowListbox: false,
    highlighted: undef,
    selected: undef,
    customStyles: styles,
    props: propsMinusChildren
  })

  useEffect(() => dispatch(setQuery(text)), [text]) // TODO: Is this needed?

  return (
    <StateContext.Provider
      value={{state, dispatch}}>
      {children}
    </StateContext.Provider>
  )
}

export { StateContext, StateContextProvider }
