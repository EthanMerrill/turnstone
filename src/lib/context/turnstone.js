import React, { createContext, useState, useEffect } from 'react'

const TurnstoneContext = createContext()

const TurnstoneContextProvider = (props) => {
  const { children, splitChar, styles = {}, text = '' } = props
  const [queryState, setQueryState] = useState(text)
  const [highlightedState, setHighlightedState] = useState()
  const [selectedState, setSelectedState] = useState()
  const [customStyles] = useState(styles)
  const [splitCharState] = useState(splitChar)

  useEffect(() => setQueryState(text), [text])

  return (
    <TurnstoneContext.Provider
      value={{
        queryState,
        setQueryState,
        highlightedState,
        setHighlightedState,
        selectedState,
        setSelectedState,
        customStyles,
        splitCharState
      }}>
      {children}
    </TurnstoneContext.Provider>
  )
}

export { TurnstoneContext, TurnstoneContextProvider }
