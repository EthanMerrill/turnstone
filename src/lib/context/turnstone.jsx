import React, { createContext, useReducer, useState, useEffect } from 'react'
import reducer from '../reducers/reducer'
import {setQuery} from '../actions/actions'
import undef from '../utils/undef'

const TurnstoneContext = createContext() //TODO: Rename GlobalStateContext

const TurnstoneContextProvider = (props) => {
  const { children, splitChar, styles = {}, text = '', items = [] } = props
  const [state, dispatch] = useReducer(reducer, {
    query: text,
    items,
    highlighted: undef,
    selected: undef,
    customStyles: styles,
    splitChar
  })
  const [selectedState, setSelectedState] = useState()
  const [customStyles] = useState(styles)
  const [splitCharState] = useState(splitChar)

  useEffect(() => dispatch(setQuery(text)), [text]) // TODO: Is this needed?

  return (
    <TurnstoneContext.Provider
      value={{
        state,
        dispatch,
        selectedState,
        setSelectedState,
        customStyles,
        splitCharState
      }}>
      {children}
    </TurnstoneContext.Provider>
  )
}

export { TurnstoneContext, TurnstoneContextProvider }
