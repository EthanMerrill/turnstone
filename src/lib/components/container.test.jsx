import React from 'react'
import renderer from 'react-test-renderer'

import { vi, describe, expect, test } from 'vitest'
import { StateContextProvider } from '../context/state'
import Container from './container'
import { fruits } from '../../data'

vi.mock('./listbox', () => ({ default: () => 'Items' }))
vi.mock('./hooks/containerEffects', () => ({
  useItemsState: vi.fn(),
  useAutoFocus: vi.fn(),
  useQueryChange: vi.fn(),
  useHighlight: vi.fn(),
  useSelected: vi.fn()
}))
vi.mock('./hooks/useData', () => ({ default: vi.fn().mockImplementation(() => [1,2,3,4,5]) }))
vi.mock('use-debounce', vi.fn().mockImplementation(() => ({
  useDebounce: (query) => [query]
})))

describe('Container', () => {
  test('Component renders correctly', () => {
    const customStyles = {
      queryContainer: 'query-container-class',
      query: 'query-class',
      typeahead: 'typeahead-class',
      x: 'x-class'
    }
    const props = {
      autoFocus: true,
      data: fruits,
      dataSearchType: 'startswith',
      debounceWait: 0,
      id: 'autocomplete',
      maxItems: 10,
      noItemsMessage: 'No matching fruit found',
      placeholder: 'Type something fruity',
      styles: customStyles
    }

    const component = renderer.create(
      <StateContextProvider {...props}>
       <Container {...props} />
      </StateContextProvider>
    )

    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})