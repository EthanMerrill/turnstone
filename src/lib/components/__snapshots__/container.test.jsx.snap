// Vitest Snapshot v1

exports[`Container > Component renders correctly 1`] = `
<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="autocomplete-listbox"
  className="query-container-class"
  role="combobox"
  style={
    {
      "display": "inline-block",
      "position": "relative",
    }
  }
>
  <input
    aria-autocomplete="both"
    aria-controls="autocomplete-listbox"
    autoCapitalize="off"
    autoComplete="off"
    autoCorrect="off"
    className="query-class"
    id="autocomplete"
    onBlur={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    placeholder="Type something fruity"
    spellCheck="false"
    style={
      {
        "backgroundColor": "transparent",
        "position": "relative",
        "zIndex": 1,
      }
    }
    type="text"
  />
  <input
    aria-hidden="true"
    autoCapitalize="off"
    autoComplete="off"
    autoCorrect="off"
    className="typeahead-class"
    readOnly="readonly"
    spellCheck="false"
    style={
      {
        "left": 0,
        "position": "absolute",
        "top": 0,
        "zIndex": 0,
      }
    }
    tabIndex="-1"
    type="text"
  />
</div>
`;
